compiler/rustc_data_structures/src/thin_vec.rs

```diff
@@ -46,6 +56,102 @@ impl<T> ::std::ops::DerefMut for ThinVec<T> {
     }
 }
 
+impl<T> FromIterator<T> for ThinVec<T> {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        let mut iter = iter.into_iter();
+
+        match iter.next() {
+            None => ThinVec(None),
+            Some(first) => {
+                // + 1 for the first element
+                let mut vec = Vec::with_capacity(iter.size_hint().0 + 1);
+
+                vec.push(first);
+                for next in iter {
+                    vec.push(next);
+                }
+
+                ThinVec(Some(Box::new(vec)))
+            }
+        }
+    }
+}
+
+impl<T> IntoIterator for ThinVec<T> {
+    type Item = T;
+    type IntoIter = IntoIterOwned<T>;
+
+    fn into_iter(self) -> IntoIterOwned<T> {
+        match self {
```

Comment:
```suggestion
        IntoIterOwned(self.0.map(|v| v.into_iter()))
```
All matches here and below should actually be replaceable with `Option`'s `map` and `and_then`.

Reply:
Outdated due to removal of custom iterators.